apps/src/code-studio/assets/SpriteLabAnimationUpload.jsx

```diff
@@ -0,0 +1,590 @@
+
+  handleCategoryChange = event => {
+    const value = event.target.value;
+    let {filename, availability, fileData} = this.state;
+    this.setState(
+      {
+        category: value,
+        metadata: '',
+        [ANIMATION_EXISTS_CHECK]: Status.WAITING
+      },
+      () => {
+        this.determineIfAnimationAlreadyExists(
+          availability,
+          filename.split('.')[0],
+          value,
+          fileData
+        );
+      }
+    );
+  };
+
+  handleAliasChange = event => {
+    const aliases = event.target.value?.split(',') || [];
+    let processedAliases = aliases.map(alias => alias.trim());
+    this.setState({aliases: processedAliases, metadata: ''});
+  };
+
+  // Set the upload status to indicate whether this file would override another
+  determineIfAnimationAlreadyExists = (
+    availability,
+    filename,
+    category,
+    fileData
+  ) => {
+    if (!fileData) {
+      return;
+    }
+    if (availability === AnimationLocations.library && category === '') {
+      this.setState({
+        [ANIMATION_EXISTS_CHECK]: Status.ALERT,
+        destination: 'No category selected'
+      });
+      return;
+    }
+    let {spriteLabLevelAnimations, spriteLabLibraryAnimations} = this.state;
+    let destination, animationExists;
+    switch (availability) {
+      case AnimationLocations.level:
+        destination = `level_animations/${filename}`;
+        animationExists = spriteLabLevelAnimations.includes(destination);
+        break;
+      case AnimationLocations.library:
+        destination = `category_${category}/${filename}`;
+        animationExists = spriteLabLibraryAnimations.includes(destination);
+        break;
+    }
+    this.setState({
+      [ANIMATION_EXISTS_CHECK]: animationExists
+        ? Status.ALERT
+        : Status.SUCCEEDED,
+      destination: destination
+    });
+  };
+
+  generateMetadata = () => {
+    const {filename, aliases, category} = this.state;
+    const metadata = {
+      name: filename.split('.')[0],
+      aliases: aliases,
+      frameCount: 1,
+      frameSize: this.state.frameSize,
+      looping: true,
+      frameDelay: 2,
+      categories: [category]
+    };
+    this.setState({metadata: JSON.stringify(metadata)});
+  };
+
+  animationExistsMessage() {
+    switch (this.state[ANIMATION_EXISTS_CHECK]) {
+      case Status.SUCCEEDED:
+        return 'This path and filename are available.';
+      case Status.ALERT:
+        return 'Filename already exists at this path. Please rename and try again.';
+      case Status.WAITING:
+        return 'Select a file and destination above.';
+    }
+  }
+  uploadStatusMessage() {
+    switch (this.state.uploadStatus) {
+      case Status.CELEBRATING:
+        return 'Animation uploaded successfully!';
+      case Status.UNKNOWN:
+        return 'Upload in progress...';
+      case Status.FAILED:
+        return this.state.errorMessage;
+      case Status.WAITING:
+        return 'Waiting...';
+    }
+  }
+
+  render() {
```

Comment:
As we are renaming the file and doing some deeper refactor work here, I think this is a good chance for us to pause and break out this mega component into smaller components. That'll make future work (and reviews!) way easier. Thoughts?

Reply:
100% agree!! My plan for now was to get this huge refactor all working (in one place), and then divide it out into smaller components as a next step. I'd love to pick your brain on the best way to do this. I think it's something that could happen on this branch before anything gets merged.

Reply:
@epeach Actually on second thought, I do wonder about possibly making this an "up next" refactor task. I know that curriculum is getting ready to upload more final HoC assets and it would be cool to have them use this new version if that were possible.